### PR TITLE
encryption: support encryption in snapshots

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Support encryption in snapshots

### What is changed and how it works?

Decrypt snapshot cf files before sending, and encrypt snapshot cf files before receiving.

### Check List 

Tests

- No code

Side effects

### Release note
Support encryption in snapshots

### Remained works
- [ ] Add test cases
- [ ] Delete useless keys in key manager

### CODE REVIEW

File: components/engine_rocks/src/import.rs [hunk 1]
```diff
@@ -44,44 +43,17 @@ impl ImportExt for RocksEngine {
...
-                .map_err(|e| format!("sync {}: {:?}", path, e))?;
-
-            let checksum = calc_crc32(path)?;
-            if checksum != expected_checksum {
-                return Err(Error::Engine(format!(
-                    "invalid checksum {} for {}, expected {}",
-                    checksum, path, expected_checksum
-                )));
-            }
-        }
```
Comment (overvenus) [inline components/engine_rocks/src/import.rs:84]: Why remove them?

Comment (hicqu) [reply components/engine_rocks/src/import.rs:84]: checksum and size are checked in `snap.rs`.

File: components/raftstore/src/store/snap.rs [hunk 2]
```diff
@@ -1224,25 +1361,39 @@ impl SnapManager {
...
             snapshot_data.take_meta(),
             Arc::clone(&core.snap_size),
             Box::new(self.clone()),
             self.limiter.clone(),
         )?;
+        f.encryption_key_manager = self.encryption_key_manager.clone();
+        if let Some(mgr) = f.encryption_key_manager.as_ref() {
+            for cf_file in &mut f.cf_files {
+                let path = cf_file.path.to_str().unwrap();
+                let enc_info = mgr.new_file(path)?;
```
Comment (yiwu-arbug) [inline components/raftstore/src/store/snap.rs:1376]: need to call `mgr.delete_file` on error cases.

File: components/tikv_util/src/file.rs [hunk 3]
```diff
@@ -76,6 +76,25 @@ pub fn calc_crc32<P: AsRef<Path>>(path: P) -> io::Result<u32> {
...
+    loop {
+        match reader.read(&mut buf[..]) {
+            Ok(0) => {
+                return Ok((digest.finalize(), fsize as u64));
+            }
+            Ok(n) => {
+                digest.update(&buf[..n]);
+                fsize += n;
+            }
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/tikv_util/src/file.rs:92]: Just like `EAGAIN or EWOULDBLOCK` for `man 2 write`.

File: components/raftstore/src/store/snap.rs [hunk 4]
```diff
@@ -821,16 +902,23 @@ impl GenericSnapshot for Snap {
             && file_exists(&self.meta_file.path)
     }
 
+    // TODO: It's very hard to handle key manager correctly without lock `SnapManager`.
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/raftstore/src/store/snap.rs:905]: The bad thing is `SnapManager`. For exmaple, it's possible that when a snapshot is applying, an another snapshot with the same name is receiving. If the second snapshot is dropped and its key is deleted from the key manager, rocksdb will not be able to ingest the first one.

So we need to simplify `SnapManager`.

**Comment by overvenus** [comment]
run-integration-br-test

File: components/raftstore/src/store/snap.rs [hunk 5]
```diff
@@ -285,17 +351,23 @@ fn check_file_size_and_checksum(
...
 
 #[derive(Default)]
-struct CfFile {
+pub struct CfFile {
     pub cf: CfName,
     pub path: PathBuf,
     pub tmp_path: PathBuf,
     pub clone_path: PathBuf,
     pub file: Option<File>,
+    pub encrypted_file: Option<DecrypterReader<File>>,
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/raftstore/src/store/snap.rs:369]: `file` is also used for saving received bytes, so we need to keep the field. but make `encrypted_file` as a box is a good idea.

**Comment by yiwu-arbug** [comment]
discussed with @tennix and @hicqu, plain file encryption (for lock CF) will be in a different PR.

File: Cargo.toml [hunk 6]
```diff
@@ -173,6 +173,9 @@ protobuf-codegen = { git = "https://github.com/pingcap/rust-protobuf", rev = "b6
 fail = { git = "https://github.com/tikv/fail-rs.git", rev = "2cf1175a1a5cc2c70bd20ebd45313afd69b558fc" }
 prometheus = { git = "https://github.com/tikv/rust-prometheus.git", rev = "a626d449eaebd5e8ce337f95c1d6dc9800f25df7" }
 
+[patch.'https://github.com/pingcap/kvproto.git']
+kvproto = { git = "https://github.com/yiwu-arbug/kvproto.git", branch = "enc_proto" }
+
```
Comment (overvenus) [inline Cargo.toml:178]: Can we remove the patch?

File: components/tikv_util/src/file.rs [hunk 7]
```diff
@@ -76,6 +76,25 @@ pub fn calc_crc32<P: AsRef<Path>>(path: P) -> io::Result<u32> {
     }
 }
 
+/// Calculates crc32 and decrypted size for a given reader.
+pub fn calc_crc32_and_size<R: Read>(reader: &mut R) -> io::Result<(u32, u64)> {
```
Comment (overvenus) [inline components/tikv_util/src/file.rs:80]: It's almost the same as the `calc_crc32` above, could you merge them together?

**Pull request merged**